### PR TITLE
fix(task): grant task-run execution role access to secrets based on tag

### PR DESCRIPTION
**Pull request by paragbhingre**
This PR fixes the issue #2641 

Fixed the issue by adding policies to the execution role when `--secrets` are provided. 

Tests - 
Following manual tests have been taken into consideration
1. When we send both `--app` and `--env` flags in `copilot task run` command then secrets are accessible only when `copilot-application` and `copilot-environment` tags are present with the correct values on the secret
2. When we send both `--app` and `--env` flags in `copilot task run` command then secrets are inaccessible only when `copilot-application` and `copilot-environment` tags are NOT present on the secret or they have incorrect values

### CODE REVIEW

File: internal/pkg/template/templates/task/cf.yml [hunk 1]
```diff
@@ -89,6 +100,22 @@ Resources:
...
         - !Sub 'arn:${AWS::Partition}:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy'
+      {{- if .Secrets}}
+      Policies:
+        - PolicyName: 'PullSecrets'
+          PolicyDocument:
+            Version: '2012-10-17'
+            Statement:
+              - Effect: 'Allow'
+                Action:
+                - ssm:GetParameters
```
Comment (Lou1415926) [inline internal/pkg/template/templates/task/cf.yml]: Should we probably also take `secretsmanager` into account 🤔

File: internal/pkg/template/templates/task/cf.yml [hunk 2]
```diff
@@ -89,6 +100,22 @@ Resources:
...
+            Version: '2012-10-17'
+            Statement:
+              - Effect: 'Allow'
+                Action:
+                - ssm:GetParameters  
+                Condition:
+                  StringEquals:
+                    "ssm:ResourceTag/copilot-application": !If [ HasAppAndEnv, !Ref App, !Ref "AWS::NoValue" ]
+                    "ssm:ResourceTag/copilot-environment": !If [ HasAppAndEnv, !Ref Env, !Ref "AWS::NoValue" ]
+                    "ssm:ResourceTag/copilot-task": !If [ HasAppAndEnv, !Ref "AWS::NoValue", !Ref TaskName ]
```
Comment (efekarakus) [inline internal/pkg/template/templates/task/cf.yml]: I really like the changes with allowing access to the `copilot-application` and `copilot-environment` tags  -- makes total sense to me. I am not sure though how obvious it's going to be when not using a copilot managed cluster, requiring the `copilot-task` tag for a one-time job seems a bit cumbersome.

What if we had an experience like this?

```
$ copilot task run --default --secrets hello=/hello,world=arn:aws:secretsmanager:us-west-2:111122223333:secret:aes128-1a2b3c
Looks like you're requesting ssm:GetParameters to the following SSM parameters:
* /hello 
and secretsmanager:GetSecretValue to the following Secrets Manager secrets:
* arn:aws:secretsmanager:us-west-2:111122223333:secret:aes128-1a2b3c

Do you grant permission to the ECS/Fargate agent for these secrets? (y/N)
```

Or if they'd like to skip the prompt they can pass a flag similar to what the CDK does with `--require-approval never`

```
$ copilot task run --default --secrets hello=/hello --require-approval never
```

Then we will augment the Task Execution role such that only the input resources are granted access.

Let me know what you think!

Comment (iamhopaul123) [reply internal/pkg/template/templates/task/cf.yml]: > Do you grant permission to the ECS/Fargate agent for these secrets? (y/N)

If they choose "N", do we just error out saying they can't use these secrets without granting access? And what are available options for `--require-approval`?

Comment (efekarakus) [reply internal/pkg/template/templates/task/cf.yml]: > If they choose "N", do we just error out saying they can't use these secrets without granting access? 

I assume we'll abort then, but we can look to what the CDK does as well.

> And what are available options for --require-approval?

hmm good point for us none, maybe we can flip the flag `--acknowledge-iam`

**Review by efekarakus — APPROVED** [review]
Looks good to me! Just to validate we tested it with both secrets (ssm and secretsmanager) and it works right? 😂 

Similarly, can we test as well with a secret that's not tagged to make sure it still cannot access it

**Comment by paragbhingre** [comment]
> Looks good to me! Just to validate we tested it with both secrets (ssm and secretsmanager) and it works right? 😂
> 
> Similarly, can we test as well with a secret that's not tagged to make sure it still cannot access it

Both the secrets (ssm and secretsmanager) and scenarios you mentioned are tested ✅

File: internal/pkg/template/templates/task/cf.yml [hunk 3]
```diff
@@ -39,6 +43,13 @@ Conditions:
     !Not [ !Equals [ !Join [ "", !Ref EntryPoint ], "" ] ]
   HasCustomPlatform:
     !Not [!Equals [!Ref OS, ""]]
+  HasApp:
+    !Not [!Equals [!Ref App, ""]]
+  HasEnv:
+    !Not [!Equals [!Ref Env, ""]]
```
Comment (iamhopaul123) [inline internal/pkg/template/templates/task/cf.yml]: Do we really need hasapp and has env?

File: internal/pkg/template/templates/task/cf.yml [hunk 4]
```diff
@@ -89,6 +100,34 @@ Resources:
...
+                  Action:
+                    - secretsmanager:GetSecretValue
+                  Condition:
+                    StringEquals:
+                      "secretsmanager:ResourceTag/copilot-application": !If [ HasAppAndEnv, !Ref App, !Ref "AWS::NoValue" ]
+                      "secretsmanager:ResourceTag/copilot-environment": !If [ HasAppAndEnv, !Ref Env, !Ref "AWS::NoValue" ]
+                  Resource:
+                    - !Sub arn:${AWS::Partition}:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:*
+          - !Ref AWS::NoValue
+      {{- end}}
```
Comment (iamhopaul123) [inline internal/pkg/template/templates/task/cf.yml]: This is awesome!

File: internal/pkg/template/templates/task/cf.yml (hunk 3 above)
Comment (paragbhingre) [reply internal/pkg/template/templates/task/cf.yml]: Actually we can remove this and put these conditions directly into `HasAppAndEnv` condition.

File: internal/pkg/template/templates/task/cf.yml [hunk 5]
```diff
@@ -89,6 +96,34 @@ Resources:
...
+            PolicyDocument:
+              Version: '2012-10-17'
+              Statement:
+                - Effect: 'Allow'
+                  Action:
+                  - ssm:GetParameters
+                  Condition:
+                    StringEquals:
+                      "ssm:ResourceTag/copilot-application": !If [ HasAppAndEnv, !Ref App, !Ref "AWS::NoValue" ]
+                      "ssm:ResourceTag/copilot-environment": !If [ HasAppAndEnv, !Ref Env, !Ref "AWS::NoValue" ]
```
Comment (Lou1415926) [inline internal/pkg/template/templates/task/cf.yml]: Can we remove these  `!If` conditionals now that we have line 101-102?

File: internal/pkg/template/templates/task/cf.yml [hunk 6]
```diff
@@ -89,6 +96,34 @@ Resources:
...
+                  Resource:
+                    - !Sub arn:${AWS::Partition}:ssm:${AWS::Region}:${AWS::AccountId}:parameter/*
+                - Effect: 'Allow'
+                  Action:
+                    - secretsmanager:GetSecretValue
+                  Condition:
+                    StringEquals:
+                      "secretsmanager:ResourceTag/copilot-application": !If [ HasAppAndEnv, !Ref App, !Ref "AWS::NoValue" ]
+                      "secretsmanager:ResourceTag/copilot-environment": !If [ HasAppAndEnv, !Ref Env, !Ref "AWS::NoValue" ]
```
Comment (Lou1415926) [inline internal/pkg/template/templates/task/cf.yml]: Samesies!

**Review by Lou1415926 — APPROVED** [review]

**Pull request merged**